mountpoint-s3/tests/reftests/reference.rs

```diff
@@ -15,62 +14,141 @@ pub enum File {
-            let mut depth = 0usize;
-            for child in map.values() {
-                depth = depth.max(1 + child.depth());
-            }
-            depth
-        } else {
-            0
-        }
-    }
 }
 
+/// The expected state of a file system. We track three pieces of state: the keys in an S3 bucket,
+/// plus lists of local files and local directories. Whenever we need the tree structure of the
+/// file system, we construct it from these inputs as a [MaterializedReference]. Building the
+/// reference in this indirect way allows us to have only one definition of correctness -- the
+/// implementation of [build_reference] -- and to test both mutations to the file system itself and
+/// "remote" mutations to the bucket (like adding or deleting a key using another client).
 #[derive(Debug)]
 pub struct Reference {
+    /// Contents of our S3 bucket
+    remote_keys: Vec<(String, MockObject)>,
+    /// Local files
+    local_files: Vec<PathBuf>,
+    /// Local directories
+    local_directories: Vec<PathBuf>,
+    /// Materialized state
+    materialized: MaterializedReference,
+}
+
+#[derive(Debug)]
+struct MaterializedReference {
     root: Node,
-    /// Full path of all directories
     directories: Vec<PathBuf>,
-    /// Full path of all inflight writes
-    #[allow(unused)] // TODO when we test partially written files
-    inflight_writes: Vec<PathBuf>,
+}
+
+impl MaterializedReference {
+    /// Add a new node to the tree. Any missing intermediate directories will be created as local
+    /// directories. If the path already exists it will be overwritten, unless both the existing
+    /// and new nodes are directories.
+    fn add_local_node(&mut self, path: impl AsRef<Path>, new_node: Node) {
+        let mut components = path.as_ref().components().peekable();
+        assert_eq!(components.next(), Some(Component::RootDir));
+
+        let mut parent_node = &mut self.root;
+        while let Some(dir) = components.next() {
+            let Node::Directory { children, .. } = parent_node else {
+                panic!("unexpected internal file node");
+            };
+            let dir = dir.as_os_str().to_str().unwrap();
+            if components.peek().is_none() {
+                // If both a local and a remote directory exist, don't overwrite the remote one's
+                // contents, as they will be visible even though the directory is local. But
+                // remember the directory is still local.
+                if let Node::Directory {
+                    children: new_children,
+                    is_local: new_is_local,
+                } = &new_node
+                {
+                    if let Some(Node::Directory {
+                        is_local: curr_is_local,
+                        ..
+                    }) = children.get_mut(dir)
+                    {
+                        assert!(new_children.is_empty(), "local directories are always empty");
```

Comment:
Is there any advantage in passing in the new_node? I suspect just passing in the type (file vs dir) would make this method simpler.

Reply:
Yeah, you're right, no idea why I tried to do it this way. I added a new `NodeType` enum instead (could have just used `fuser::FileType` directory but it has extra variants we don't allow).